File: core/src/main/scala/com/graphbrain/eco/nodes/FilterFun.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.{Context, Contexts, NodeType}

class FilterFun(val fun: FilterFun.FilterFun, params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {
  override val label = fun match {
    case FilterFun.FilterMin => "filter-min"
    case FilterFun.FilterMax => "filter-max"
  }

  override def ntype: NodeType = NodeType.Boolean

  override def booleanValue(ctxts: Contexts) = {
    params(0).numberValue(ctxts)

    var bestCtxt: Context = null
    var bestVal = fun match {
      case FilterFun.FilterMin => Double.PositiveInfinity
      case FilterFun.FilterMin => Double.NegativeInfinity
    }

    for (c <- ctxts.ctxts) {
      val value = c.getRetNumber(params(0))

      fun match {
        case FilterFun.FilterMin =>
          if (value < bestVal) {
            bestCtxt = c
            bestVal = value
          }
        case FilterFun.FilterMax =>
          if (value > bestVal) {
            bestCtxt = c
            bestVal = value
          }
      }
    }

    for (c <- ctxts.ctxts)
      c.setRetBoolean(this, c eq bestCtxt)
  }

  override protected def typeError() = error("parameters must be boolean")
}

object FilterFun extends Enumeration {
  type FilterFun = Value
  val FilterMin,
  FilterMax = Value
}

File: core/src/main/java/com/graphbrain/eco/nodes/FilterFun.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Context;
import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

class FilterFun extends FunNode {

    enum FilterFunType {
        FilterMin, FilterMax
    }

    private FilterFunType fun;

    public FilterFun(FilterFunType fun, ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
        this.fun = fun;
    }

    public FilterFun(FilterFunType fun, ProgNode[] params) {
        this(fun, params, -1);
    }

    @Override
    public String label() {
        switch(fun) {
            case FilterMin: return "filter-min";
            case FilterMax: return "filter-max";
        }
        // error
        return "?";
    }

    @Override
    public NodeType ntype(){return NodeType.Boolean;}

    @Override
    public void booleanValue(Contexts ctxts) {
        params[0].numberValue(ctxts);

        Context bestCtxt = null;

        double bestVal = 0;
        switch(fun) {
            case FilterMin:
                bestVal = Double.POSITIVE_INFINITY;
                break;
            case FilterMax:
                bestVal = Double.NEGATIVE_INFINITY;
                break;
        }

        for (Context c : ctxts.getCtxts()) {
            double value = c.getRetNumber(params[0]);

            switch(fun) {
                case FilterMin:
                    if (value < bestVal) {
                        bestCtxt = c;
                        bestVal = value;
                    }
                    break;
                case FilterMax:
                    if (value > bestVal) {
                        bestCtxt = c;
                        bestVal = value;
                    }
                    break;
            }
        }

        for (Context c : ctxts.getCtxts()) {
            if (bestCtxt == null) {
                c.setRetBoolean(this, false);
            }
            else {
                c.setRetBoolean(this, c == bestCtxt);
            }
        }
    }
}